### PR TITLE
[4.1] Set a header limit for server timing

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #37080.

### Summary of Changes
Define a header limit for the server response time in debug mode. This is required due some limits in web servers.

The limit is set to 4000 which should fulfill all common web servers. according to https://stackoverflow.com/a/8623061.

### Testing Instructions
- Use nginx web server.
- Create a lot of dashboard modules.
- Enable debug mode in Joomla configuration.

### Actual result BEFORE applying this Pull Request
Bad gatweway error as in issue.

### Expected result AFTER applying this Pull Request
Site is rendered properly.

### CODE REVIEW

**Comment by zero-24** [comment]
Can we get a backport of this for 3.10 too?

**Comment by Fedik** [comment]
Maybe a better do not add $metric when it too big, because adding "half data" will show incorrect "timing", that may mislead.
Or I missed something?

**Comment by Fedik** [comment]
@zero-24 we do not have it in 3.10, it was a new feature in 4.1 :wink:

**Comment by zero-24** [comment]
Ha ok no backport needed in this case :D

**Comment by richard67** [comment]
@laoneo Does this PR solve issue #37080 so the issue can be closed?

**Comment by laoneo** [comment]
> @laoneo Does this PR solve issue #37080 so the issue can be closed?

At the end of the day it should solve it

**Comment by laoneo** [comment]
> Maybe a better do not add $metric when it too big, because adding "half data" will show incorrect "timing", that may mislead. Or I missed something?

I was thinking the same, perhaps we should then return one entry with a message too much data.

**Comment by mikeazores** [comment]
In reference to #37080 if I lower the strlen value to say 3600 IT WORKS. With the current PR value (4000) I still have problems.

**Comment by zero-24** [comment]
@laoneo do you think it makes sense to make that a option within the plugin with a default value for 4k (Yes I also hate new options)? So on sites such as @mikeazores 's he can lower the value but on other sites we can use the 4k / 8k etc.?

**Comment by Fedik** [comment]
> I was thinking the same, perhaps we should then return one entry with a message too much data.

Or maybe do only 2 point timing: between start-end (when too big).
Sadly we cannot easily exclude timings for modules.

**Comment by laoneo** [comment]
> @laoneo do you think it makes sense to make that a option within the plugin with a default value for 4k (Yes I also hate new options)? So on sites such as @mikeazores 's he can lower the value but on other sites we can use the 4k / 8k etc.?

For me it should work properly out of the box and it is not that critical, that it must always return a value.



> > I was thinking the same, perhaps we should then return one entry with a message too much data.
> 
> Or maybe do only 2 point timing: between start-end (when too big). Sadly we cannot easily exclude timings for modules.

We can, and deliver module rendering as one metric when there is a string like `mod_` in the profile step.

**Comment by BrainforgeUK** [comment]
Sorry, 4000 limit is too large, I had to make it 3500 to get it to work on my server.
See https://github.com/joomla/joomla-cms/issues/37090

The server-wide Nginx settings for my server are 'out of the box'.

Can you making it configurable on the debug plugin parameters?
Default to 2000?
Warning about needing to change nginx settings if you increase it?
Suggest those nginx settings I grabbed earlier as examples?

**Comment by laoneo** [comment]
@BrainforgeUK probably the better approach would be to collect the module ones as one entry. Having an option for such a small feature would be overhead.

**Comment by laoneo** [comment]
The last commit adds a new behavior where the modules are collected together and are printed as one entry. @Fedik what do you think?

It looks then like:

image

**Comment by Fedik** [comment]
Looks good to me.

Additionally still better to keep your previous approach with limit check, just to be safe, around 3K.
Because we cannot guarantee size when any extensions will add own Profiling data.

**Comment by laoneo** [comment]
Added a limit again. Can you guys test it now so we can ship it with the next patch release?

File: plugins/system/debug/debug.php [hunk 1]
```diff
@@ -691,13 +692,28 @@ public function onBeforeRespond(): void
...
 
+				continue;
+			}
+
+			$desc     = str_ireplace('after', '', $mark->label);
 			$name     = preg_replace('/[^\da-z]/i', '', $desc);
-			$metrics .= sprintf('%s;dur=%f;desc="%s:", ', $index . $name, $mark->time, $desc);
+			$metrics .= sprintf('%s;dur=%f;desc="%s", ', $index . $name, $mark->time, $desc);
+
+			// Do not crete too large headers, some web servers don't love them
```
Comment (Quy) [inline plugins/system/debug/debug.php]:
```suggestion
			// Do not create too large headers, some web servers don't love them
```

**Comment by HLeithner** [comment]
Can you Plesse add 2 things?
1. make the limit a parameter
2. Add a line that to the metric that the limit got hit.

So it's possible to know that something is missing and that you can adjust it depending on your needs.

**Comment by laoneo** [comment]
Adding  a new parameter for such a small feature makes no sense and the majority will not even understand for what it is.

**Comment by HLeithner** [comment]
The parameter is in the debug plugin, the person configuring the plugin should know what the web does but OK, it's not important I guess. But the information that the data is truncated is important from my point of view.

**Comment by laoneo** [comment]
How would you render that message?

**Comment by Fedik** [comment]
> How would you render that message?

Kind of last row in metrics
```php
$metrics .= sprintf('%s;dur=%f;desc="%s", ', $index . 'Rest of data is gone forever', 'infinity', '');
```
:smiley:

**Comment by laoneo** [comment]
Last commit shows now a truncated message:

image

If you want to test it, change the code `if (strlen($metrics) > 3000)` to `if (strlen($metrics) > 1)`.

File: plugins/system/debug/debug.php [hunk 2]
```diff
@@ -691,13 +692,29 @@ public function onBeforeRespond(): void
 				continue;
 			}
 
-			$desc = $mark->label;
-			$desc = str_ireplace('after', '', $desc);
+			// Collect the module render time
+			if (strpos($mark->label, 'mod_') !== false)
+			{
+				$moduleTime += $mark->time;
```
Comment (Quy) [inline plugins/system/debug/debug.php:702]:
```suggestion
```

File: plugins/system/debug/debug.php [hunk 3]
```diff
@@ -691,13 +692,29 @@ public function onBeforeRespond(): void
...
+
+			$desc     = str_ireplace('after', '', $mark->label);
 			$name     = preg_replace('/[^\da-z]/i', '', $desc);
-			$metrics .= sprintf('%s;dur=%f;desc="%s:", ', $index . $name, $mark->time, $desc);
+			$metrics .= sprintf('%s;dur=%f;desc="%s", ', $index . $name, $mark->time, $desc);
+
+			// Do not create too large headers, some web servers don't love them
+			if (strlen($metrics) > 3000)
+			{
+				$metrics .= 'System;dur=0;desc="Data truncated to 3000 characters"';
```
Comment (Quy) [inline plugins/system/debug/debug.php]:
```suggestion
				$metrics .= 'System;dur=0;desc="Data truncated to 3000 characters", ';
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 9767226fbdde11c0cf3e7ae2f454139fc38e3672<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by laoneo** [comment]
@BrainforgeUK and @mikeazores can you test it again as well?

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 9767226fbdde11c0cf3e7ae2f454139fc38e3672<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by richard67** [comment]
Back to pending due to new changes.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by laoneo** [comment]
We had the same problem with access when a none admin get logged in on the back end. The recent changes are grouping also access.

**Comment by HusiBT** [comment]
I have tested this item :white_check_mark: successfully on 105dc3d82423f673f1ed72a69444695f9536696b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by HLeithner** [comment]
I have tested this item :white_check_mark: successfully on 105dc3d82423f673f1ed72a69444695f9536696b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37087">issues.joomla.org/tracker/joomla-cms/37087</a>.</sub>

File: plugins/system/debug/debug.php [hunk 4]
```diff
@@ -691,13 +693,38 @@ public function onBeforeRespond(): void
...
-			$desc = $mark->label;
-			$desc = str_ireplace('after', '', $desc);
+			// Collect the module render time
+			if (strpos($mark->label, 'mod_') !== false)
+			{
+				$moduleTime += $mark->time;
+				continue;
+			}
+
+			// Collect the module render time
```
Comment (Quy) [inline plugins/system/debug/debug.php]:
```suggestion
			// Collect the access render time
```

**Comment by bembelimen** [comment]
Can't we just put the array in a variable, count and slice out the items which are too much from the middle (https://www.php.net/manual/en/function.array-splice.php) so we have a start/end time?

**Comment by laoneo** [comment]
You just have a time of a task, there is no start/end time. This is not a whaterfall chart.

**Comment by laoneo** [comment]
I'm setting this one as a release blocker as it affects a lot of sites and should definitely be fixed and finally shipped with 4.1.1.

**Comment by richard67** [comment]
@laoneo Well, the last comment was from @bembelimen , so it would be good to know if he is ok with your reply to it,

**Comment by laoneo** [comment]
He didn't reply so my assumption was that he did understand my explanation. I mean the pr is so correct, if he wants to make some adjustments, then he can do that in a followup pr.

**Comment by richard67** [comment]
Sure.

**Comment by bembelimen** [comment]
Thx

**Pull request merged**